Update libedgetpu repo for compatibility of recent versions of Tensorflow.

feranick opened the pull request:
This PR:
- Updates the URL calls to pull recent versions of TF (currently `v2.15.0`), TF/crosstools, and libusb (1.0.26).
- Updates versions of bazel needed in docker.
- It refactors WORKSPACE to conform to the deprecation of TensorFlow/Toolchains separate repo. [More info here.](https://discuss.tensorflow.org/t/tensorflow-toolchains-has-been-deprecated-and-moved-into-tensorflow-tensorflow/7713)
- Use xz compression for compatibility with older versions of debian (bullseye)
- Adds missing support for "stddef.h"

This is in relation to issues: https://github.com/tensorflow/tensorflow/issues/62371 and https://github.com/google-coral/libedgetpu/issues/53

feranick committed:
Add support for "stddef.h"

Prevent compilation fail

feranick committed:
UPdated README.md

feranick committed:
Update version of TF to 2.5.3

feranick committed:
Fix typos

feranick committed:
Build against TF v2.8.4

feranick committed:
Build against TF2.9.3

feranick committed:
Update crosstools

feranick committed:
Updated to support TF 2.10.1

feranick committed:
Added support for TF 2.11.1

feranick committed:
Added support for TF2.12.1

feranick committed:
Added support for TF 2.13.1

feranick committed:
Added repos for TF 2.14.1 and TF2.15.0 - not working yet.

feranick committed:
fix typo

feranick committed:
Added notes on compilation under MacOS

feranick committed:
Added Linux dependencies to python3-dev

feranick committed:
Update version of Bazeel in Dockerfile.

feranick committed:
Update README with more modern images for docker

feranick committed:
Notes to build dbian debs from docker builds

feranick committed:
Added custom dbian/rules for docker buiilds

feranick committed:
Preparation for making deb packages via Docker

feranick committed:
Revert commit 93d8e41

No longer needed. deb packaging will be done natively.

feranick committed:
Add notes on how to package debs for arm64 and armhf

feranick committed:
Use xz for deb compression for compatibility with debian

feranick committed:
Fix indentation

feranick committed:
Updated README.md

feranick committed:
Updated libusb to version 1.0.26

feranick committed:
Updated debian/changelog

feranick committed:
Added support for TensorFlow v2.15.0

Refactored WORKSPACE after deprecation of Tensorflow/toolchains. More info:
https://discuss.tensorflow.org/t/tensorflow-toolchains-has-been-deprecated-and-moved-into-tensorflow-tensorflow/7713

feranick committed:
Updated required version of bazel for Docker

Skillnoob commented:
@dmitriykovalev can you or some other maintainer please review this pr. Merging this would fix many of the issues users of the coral edge tpu have with newer tensorflow versions and python versions.

cocoa-xu commented:
> @dmitriykovalev can you or some other maintainer please review this pr. Merging this would fix many of the issues users of the coral edge tpu have with newer tensorflow versions and python versions.

This repo seems to be not maintained for a long time, and I have to create my own fork and maintain the code [there](https://github.com/cocoa-xu/libedgetpu). In that repo, I've also fixed the support for armv6 devices, which I can create another PR.

feranick committed:
Updated libusb binary for Windows to version 1.0.27.

feranick committed:
Update README.md

stakach commented:
this is awesome. @dmitriykovalev @PeterMalkin maybe allow the community to take over maintenance if Google is no longer interested?

Skillnoob commented:
> this is awesome. @dmitriykovalev @PeterMalkin maybe allow the community to take over maintenance if Google is no longer interested?

I fully agree. They should find someone who is interested in maintaining the project and can be trusted to not do anything malicious.

feranick committed:
Updated README.md

feranick committed:
Update Dockerfile.wndows

feranick committed:
Updated README.md

feranick commented:
Copying @pkgoogle, hoping this can move forward.

feranick committed:
Added support for TF 2.16.0-rc0

feranick committed:
Fix checksum for TF 2.16.0-rc0

feranick committed:
Fix wrong version and checksum for crosstool

Regresson introduced by mistake in previous commits

pkgoogle commented:
I don't have access to this repo, but I'll see if I can help bring some attention to this.

feranick commented:
Thanks, @pkgoogle, much appreciated.

jakew009 commented:
> I don't have access to this repo, but I'll see if I can help bring some attention to this.

I think lots of people would be hugely grateful if you could.

There are lots of people using this in commercial projects with the chips soldered to PCB that really are being left in the lurch :(

pkgoogle commented:
Hi @jakew009, Thanks for the information -- if you have any data on how many people this is affecting (besides what I can gather from this thread) I can push harder on my end.

feranick commented:
Hi @pkgoogle, this is just to give you a bit more context. To function, 
edgetpu needs both the libedgetpu library (and some functionality is 
only available with the optional pycoral/libcoral libraries). the 
libedgetpu library was last updated years ago, and it only supports old 
versions of TF (2.5) and Python. AS most platform (OS, python3, etc) 
have moved on, getting machines to use libedgetpu is basically 
impossible and with it the actual use of the edgetpu itself. For 
example, Raspberry Pi 4/5 ships with Python3.11, yet there is no support 
for it.

One may say: "well, the repo is there, compile your own libraries". 
Unfortunately, that won't work as the repo needs updates to make it 
viable and in sync with current TF and Python (and OS). This is what my 
PR provides.

Furthermore: One would understand that if Coral EdgeTPU were 
discontinued, than lack of support would be frustrating (as 
discontinuing sales should not mean discontinuing support) but a fact of 
life. And yet, not only EdgeTPU is not discontinued, but pretty much 
still sold by several channels (including the official coral.ai). In 
fact, I have email exchange with the sales team for the EdgeTPU platform 
(which is operated by ASUS), that confirms that EdgeTPU is absolutely 
not discontinued and according to them fully supported by the google 
team. This is of course not true, as I gather, but it shows a 
discontinuity in how the platform is managed. In essence this is is TPU 
you can buy (even through professional bulk sales) but may not be able 
to run it anywhere.

Besides, in the age where every company is launching their NPUs with 
grand fanfare, it's sad to see a pioneer gadget left with no support. 
People may see it as a a great opportunity to get one only to be left 
with an useless device.

I hope this helps. Thanks very much, your push is much appreciated.

Nicola

P.S. I always thought the edgeTPU would have made it a great AI chip 
for, say, a chromebook. Too visionary, I guess.



On 2/28/24 2:00 PM, pkgoogle wrote:
>
> Hi @jakew009 <https://github.com/jakew009>, Thanks for the information 
> -- if you have any data on how many people this is affecting (besides 
> what I can gather from this thread) I can push harder on my end.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1969652135>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAAIY7ZQZLCZF3DWARRPGHTYV55EXAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRZGY2TEMJTGU>.
> You are receiving this because you commented.Message ID: 
> ***@***.***>
>

--------------7sWOA1IvVmq4dprI7CU0c5G5
Content-Type: text/html; charset=UTF-8
Content-Transfer-Encoding: 8bit

<!DOCTYPE html><html><head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  </head>
  <body>
    Hi @pkgoogle, this is just to give you a bit more context. To
    function, edgetpu needs both the libedgetpu library (and some
    functionality is only available with the optional pycoral/libcoral
    libraries). the libedgetpu library was last updated years ago, and
    it only supports old versions of TF (2.5) and Python. AS most
    platform (OS, python3, etc) have moved on, getting machines to use
    libedgetpu is basically impossible and with it the actual use of the
    edgetpu itself. For example, Raspberry Pi 4/5 ships with Python3.11,
    yet there is no support for it. <br>
    <br>
    One may say: &quot;well, the repo is there, compile your own libraries&quot;.
    Unfortunately, that won't work as the repo needs updates to make it
    viable and in sync with current TF and Python (and OS). This is what
    my PR provides. <br>
    <br>
    Furthermore: One would understand that if Coral EdgeTPU were
    discontinued, than lack of support would be frustrating (as
    discontinuing sales should not mean discontinuing support) but a
    fact of life. And yet, not only EdgeTPU is not discontinued, but
    pretty much still sold by several channels (including the official
    coral.ai). In fact, I have email exchange with the sales team for
    the EdgeTPU platform (which is operated by ASUS), that confirms that
    EdgeTPU is absolutely not discontinued and according to them fully
    supported by the google team. This is of course not true, as I
    gather, but it shows a discontinuity in how the platform is managed.
    In essence this is is TPU you can buy (even through professional
    bulk sales) but may not be able to run it anywhere.<br>
    <br>
    Besides, in the age where every company is launching their NPUs with
    grand fanfare, it's sad to see a pioneer gadget left with no
    support. People may see it as a a great opportunity to get one only
    to be left with an useless device. <br>
    <br>
    I hope this helps. Thanks very much, your push is much appreciated.<br>
    <br>
    Nicola<br>
    <br>
    P.S. I always thought the edgeTPU would have made it a great AI chip
    for, say, a chromebook. Too visionary, I guess. <br>
    <br>
    <br>
    <br>
    <div class="moz-cite-prefix">On 2/28/24 2:00 PM, pkgoogle wrote:<br>
    </div>
    <blockquote type="cite" ***@***.***">
      
      <p dir="auto">Hi <a class="user-mention notranslate" data-hovercard-type="user" data-hovercard-url="/users/jakew009/hovercard" data-octo-click="hovercard-link-click" data-octo-dimensions="link_type:self" href="https://github.com/jakew009" ***@***.***</a>,
        Thanks for the information -- if you have any data on how many
        people this is affecting (besides what I can gather from this
        thread) I can push harder on my end.</p>
      <p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">—<br>
        Reply to this email directly, <a href="https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1969652135" moz-do-not-send="true">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AAAIY7ZQZLCZF3DWARRPGHTYV55EXAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRZGY2TEMJTGU" moz-do-not-send="true">unsubscribe</a>.<br>
        You are receiving this because you commented.<img src="https://github.com/notifications/beacon/AAAIY76EA23X6KQYC4DAGVLYV55EXA5CNFSM6AAAAABCTNMRJ2WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTVM2A2O.gif" alt="" moz-do-not-send="true" width="1" height="1"><span style="color: transparent; font-size: 0; display: none; visibility: hidden; overflow: hidden; opacity: 0; width: 0; height: 0; max-width: 0; max-height: 0; mso-hide: all">Message
          ID: <span>&lt;google-coral/libedgetpu/pull/60/c1969652135</span><span>@</span><span>github</span><span>.</span><span>com&gt;</span></span></p>
      <script type="application/ld+json">[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1969652135",
"url": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1969652135",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]</script>
    </blockquote>
    <br>
  </body>
</html>

--------------7sWOA1IvVmq4dprI7CU0c5G5--

Skillnoob commented:
@feranick I have recently written a guide on how to get the usb accelerator working with ultralytics yolov8 based on your updated libs which can be found [here](https://docs.ultralytics.com/guides/coral-edge-tpu-on-raspberry-pi/). This may be helpful for people trying to get the tpu working.

jakew009 commented:
> Hi @jakew009, Thanks for the information -- if you have any data on how many people this is affecting (besides what I can gather from this thread) I can push harder on my end.

Thanks for replying.

I can't answer for other people, but for just us, we have about 1200 Google Corals in production, and a further 1000 or so in stock (because we had to forward buy 2 years in advance, because they were so hard to get hold of). These are the units that are physically soldered onto our PCBs that we sell to customers :)

It's really scary for a small company to have a chip soldered onto PCBs that we have invested hundreds of thousands of into, and to be faced with the possibility that Google might completed abandon the software required to make them work. 

I accept that Google are no longer developing new Coral products, but it doesn't seem too much to ask that they at least maintain the software required to make these chips work.

I'm also happy to pass on the email exchanges we have had with the Asus support team (who seem to have no idea what is going on other than saying that they cannot get hold of anyone at Google either). My email is jake@ruggednetworks.co.uk.

Namburger commented on workspace.bzl:
```diff
@@ -6,11 +6,55 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# TF release 2.15.0 as of 11/10/2023. Current
+#TENSORFLOW_COMMIT = "6887368d6d46223f460358323c4b76d61d1558a8"
+#TENSORFLOW_SHA256 = "bb25fa4574e42ea4d452979e1d2ba3b86b39569d6b8106a846a238b880d73652"
```
thanks for keeping record of old tf versions and dates!

feranick commented:
@Namburger Thank you. The fork with this PR may have some redundancies, but I am happy to align it with Google guidelines. Just let me know. 

BTW, as you also noticed, the current head is built against the upcoming TF 2.16.0 (currently in rc0). I will keep updating it until it reaches stable, yet you may find a libedgetpu "stable" release in here:

https://github.com/feranick/libedgetpu/releases/tag/v16.0-TF2.15.0-1

Namburger commented on README.md:
````diff
@@ -8,29 +8,48 @@ This software is distributed in the binary form at [coral.ai/software](https://c
 $ make
 ```
 
+Rquired libraries for Linux:
````
"Required"

feranick committed:
Fixed typo.

Thank you @Numburger

Reference: https://github.com/google-coral/libedgetpu/pull/60

Namburger commented:
@feranick thank you for your contribution and keeping the project updated.

I haven't been involved with google-coral for a while now, however I just regained access to maintain this repo and I'd like to commit to keeping it healthy. I'm not sure if I'll be able to commit code changes, however I can help merge changes that are necessary to keep this project healthy and updated.

Namburger commented:
> @Namburger Thank you. The fork with this PR may have some redundancies, but I am happy to align it with Google guidelines. Just let me know.
> 
> BTW, as you also noticed, the current head is built against the upcoming TF 2.16.0 (currently in rc0). I will keep updating it until it reaches stable, yet you may find a libedgetpu "stable" release in here:
> 
> https://github.com/feranick/libedgetpu/releases/tag/v16.0-TF2.15.0-1

@feranick Do you think it make sense for this PR to be on a stable release and later you can submit another PR once we have a stable TF 2.16.0?

feranick commented:
Thanks again, @namburger, this is great news.

I am happy to help. Ideally some code changes are needed (in the PR) to 
make the library viable going forward. I am sure you can cherry pick 
commits from the PR to get there.

On 2/29/24 3:45 PM, Nam Vu wrote:
>
> @feranick <https://github.com/feranick> thank you for your 
> contribution and keeping the project updated.
>
> I haven't been involved with google-coral for a while now, however I 
> just regained access to maintain this repo and I'd like to commit to 
> keeping it healthy. I'm not sure if I'll be able to commit code 
> changes, however I can help merge changes that are necessary to keep 
> this project healthy and updated.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971929570>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAAIY7ZOTYIKWWRXGL3HPKLYV6JO7AVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZRHEZDSNJXGA>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

--------------Gwrl20FRUj1VoYkm8hpqWRso
Content-Type: text/html; charset=UTF-8
Content-Transfer-Encoding: 8bit

<!DOCTYPE html><html><head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  </head>
  <body>
    Thanks again, @namburger, this is great news. <br>
    <br>
    I am happy to help. Ideally some code changes are needed (in the PR)
    to make the library viable going forward. I am sure you can cherry
    pick commits from the PR to get there. <br>
    <br>
    <div class="moz-cite-prefix">On 2/29/24 3:45 PM, Nam Vu wrote:<br>
    </div>
    <blockquote type="cite" ***@***.***">
      
      <p dir="auto"><a class="user-mention notranslate" data-hovercard-type="user" data-hovercard-url="/users/feranick/hovercard" data-octo-click="hovercard-link-click" data-octo-dimensions="link_type:self" href="https://github.com/feranick" ***@***.***</a>
        thank you for your contribution and keeping the project updated.</p>
      <p dir="auto">I haven't been involved with google-coral for a
        while now, however I just regained access to maintain this repo
        and I'd like to commit to keeping it healthy. I'm not sure if
        I'll be able to commit code changes, however I can help merge
        changes that are necessary to keep this project healthy and
        updated.</p>
      <p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">—<br>
        Reply to this email directly, <a href="https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971929570" moz-do-not-send="true">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AAAIY7ZOTYIKWWRXGL3HPKLYV6JO7AVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZRHEZDSNJXGA" moz-do-not-send="true">unsubscribe</a>.<br>
        You are receiving this because you were mentioned.<img src="https://github.com/notifications/beacon/AAAIY775US6DOD4PDJD6MM3YV6JO7A5CNFSM6AAAAABCTNMRJ2WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTVRFA6E.gif" alt="" moz-do-not-send="true" width="1" height="1"><span style="color: transparent; font-size: 0; display: none; visibility: hidden; overflow: hidden; opacity: 0; width: 0; height: 0; max-width: 0; max-height: 0; mso-hide: all">Message
          ID: <span>&lt;google-coral/libedgetpu/pull/60/c1971929570</span><span>@</span><span>github</span><span>.</span><span>com&gt;</span></span></p>
      <script type="application/ld+json">[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971929570",
"url": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971929570",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]</script>
    </blockquote>
    <br>
  </body>
</html>

--------------Gwrl20FRUj1VoYkm8hpqWRso--

feranick committed:
Revert PR to last stable release TF 2.15.0

feranick commented:
Absolutely. I just reverted back to TF 2.15.0 in the last commit.

On 2/29/24 3:48 PM, Nam Vu wrote:
>
>     @Namburger <https://github.com/Namburger> Thank you. The fork with
>     this PR may have some redundancies, but I am happy to align it
>     with Google guidelines. Just let me know.
>
>     BTW, as you also noticed, the current head is built against the
>     upcoming TF 2.16.0 (currently in rc0). I will keep updating it
>     until it reaches stable, yet you may find a libedgetpu "stable"
>     release in here:
>
>     https://github.com/feranick/libedgetpu/releases/tag/v16.0-TF2.15.0-1
>
> @feranick <https://github.com/feranick> Do you think it make sense for 
> this PR to be on a stable release and later you can submit another PR 
> once we have a stable TF 2.16.0?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971934031>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAAIY75Q4ISECHAXSUQDWVDYV6J3DAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZRHEZTIMBTGE>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

--------------TgS00A6MmYmDkS5QEFmWVzWl
Content-Type: text/html; charset=UTF-8
Content-Transfer-Encoding: 8bit

<!DOCTYPE html><html><head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  </head>
  <body>
    Absolutely. I just reverted back to TF 2.15.0 in the last commit.<br>
    <br>
    <div class="moz-cite-prefix">On 2/29/24 3:48 PM, Nam Vu wrote:<br>
    </div>
    <blockquote type="cite" ***@***.***">
      
      <blockquote>
        <p dir="auto"><a class="user-mention notranslate" data-hovercard-type="user" data-hovercard-url="/users/Namburger/hovercard" data-octo-click="hovercard-link-click" data-octo-dimensions="link_type:self" href="https://github.com/Namburger" ***@***.***</a>
          Thank you. The fork with this PR may have some redundancies,
          but I am happy to align it with Google guidelines. Just let me
          know.</p>
        <p dir="auto">BTW, as you also noticed, the current head is
          built against the upcoming TF 2.16.0 (currently in rc0). I
          will keep updating it until it reaches stable, yet you may
          find a libedgetpu &quot;stable&quot; release in here:</p>
        <p dir="auto"><a href="https://github.com/feranick/libedgetpu/releases/tag/v16.0-TF2.15.0-1" moz-do-not-send="true" class="moz-txt-link-freetext">https://github.com/feranick/libedgetpu/releases/tag/v16.0-TF2.15.0-1</a></p>
      </blockquote>
      <p dir="auto"><a class="user-mention notranslate" data-hovercard-type="user" data-hovercard-url="/users/feranick/hovercard" data-octo-click="hovercard-link-click" data-octo-dimensions="link_type:self" href="https://github.com/feranick" ***@***.***</a>
        Do you think it make sense for this PR to be on a stable release
        and later you can submit another PR once we have a stable TF
        2.16.0?</p>
      <p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">—<br>
        Reply to this email directly, <a href="https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971934031" moz-do-not-send="true">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AAAIY75Q4ISECHAXSUQDWVDYV6J3DAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZRHEZTIMBTGE" moz-do-not-send="true">unsubscribe</a>.<br>
        You are receiving this because you were mentioned.<img src="https://github.com/notifications/beacon/AAAIY7YAGRLND6DMRQH4LSTYV6J3DA5CNFSM6AAAAABCTNMRJ2WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTVRFJU6.gif" alt="" moz-do-not-send="true" width="1" height="1"><span style="color: transparent; font-size: 0; display: none; visibility: hidden; overflow: hidden; opacity: 0; width: 0; height: 0; max-width: 0; max-height: 0; mso-hide: all">Message
          ID: <span>&lt;google-coral/libedgetpu/pull/60/c1971934031</span><span>@</span><span>github</span><span>.</span><span>com&gt;</span></span></p>
      <script type="application/ld+json">[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971934031",
"url": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971934031",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]</script>
    </blockquote>
    <br>
  </body>
</html>

--------------TgS00A6MmYmDkS5QEFmWVzWl--

feranick committed:
Revert to bazel 6.1.0 for for Docker.windows for TF 2.15.0

feranick commented:
The PR should now be in the stable release as per commit:

https://github.com/feranick/libedgetpu/commit/d90ba0125d28b5288a14eb67b1ebaeeaeae5b37a

On 2/29/24 3:48 PM, Nam Vu wrote:
>
>     @Namburger <https://github.com/Namburger> Thank you. The fork with
>     this PR may have some redundancies, but I am happy to align it
>     with Google guidelines. Just let me know.
>
>     BTW, as you also noticed, the current head is built against the
>     upcoming TF 2.16.0 (currently in rc0). I will keep updating it
>     until it reaches stable, yet you may find a libedgetpu "stable"
>     release in here:
>
>     https://github.com/feranick/libedgetpu/releases/tag/v16.0-TF2.15.0-1
>
> @feranick <https://github.com/feranick> Do you think it make sense for 
> this PR to be on a stable release and later you can submit another PR 
> once we have a stable TF 2.16.0?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971934031>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAAIY75Q4ISECHAXSUQDWVDYV6J3DAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZRHEZTIMBTGE>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

--------------vLrwj5qd3UAOZyER8PRUjCWY
Content-Type: text/html; charset=UTF-8
Content-Transfer-Encoding: 8bit

<!DOCTYPE html><html><head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  </head>
  <body>
    The PR should now be in the stable release as per commit:<br>
    <br>
<a class="moz-txt-link-freetext" href="https://github.com/feranick/libedgetpu/commit/d90ba0125d28b5288a14eb67b1ebaeeaeae5b37a">https://github.com/feranick/libedgetpu/commit/d90ba0125d28b5288a14eb67b1ebaeeaeae5b37a</a><br>
    <br>
    <div class="moz-cite-prefix">On 2/29/24 3:48 PM, Nam Vu wrote:<br>
    </div>
    <blockquote type="cite" ***@***.***">
      
      <blockquote>
        <p dir="auto"><a class="user-mention notranslate" data-hovercard-type="user" data-hovercard-url="/users/Namburger/hovercard" data-octo-click="hovercard-link-click" data-octo-dimensions="link_type:self" href="https://github.com/Namburger" ***@***.***</a>
          Thank you. The fork with this PR may have some redundancies,
          but I am happy to align it with Google guidelines. Just let me
          know.</p>
        <p dir="auto">BTW, as you also noticed, the current head is
          built against the upcoming TF 2.16.0 (currently in rc0). I
          will keep updating it until it reaches stable, yet you may
          find a libedgetpu &quot;stable&quot; release in here:</p>
        <p dir="auto"><a href="https://github.com/feranick/libedgetpu/releases/tag/v16.0-TF2.15.0-1" moz-do-not-send="true" class="moz-txt-link-freetext">https://github.com/feranick/libedgetpu/releases/tag/v16.0-TF2.15.0-1</a></p>
      </blockquote>
      <p dir="auto"><a class="user-mention notranslate" data-hovercard-type="user" data-hovercard-url="/users/feranick/hovercard" data-octo-click="hovercard-link-click" data-octo-dimensions="link_type:self" href="https://github.com/feranick" ***@***.***</a>
        Do you think it make sense for this PR to be on a stable release
        and later you can submit another PR once we have a stable TF
        2.16.0?</p>
      <p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">—<br>
        Reply to this email directly, <a href="https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971934031" moz-do-not-send="true">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AAAIY75Q4ISECHAXSUQDWVDYV6J3DAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZRHEZTIMBTGE" moz-do-not-send="true">unsubscribe</a>.<br>
        You are receiving this because you were mentioned.<img src="https://github.com/notifications/beacon/AAAIY7YAGRLND6DMRQH4LSTYV6J3DA5CNFSM6AAAAABCTNMRJ2WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTVRFJU6.gif" alt="" moz-do-not-send="true" width="1" height="1"><span style="color: transparent; font-size: 0; display: none; visibility: hidden; overflow: hidden; opacity: 0; width: 0; height: 0; max-width: 0; max-height: 0; mso-hide: all">Message
          ID: <span>&lt;google-coral/libedgetpu/pull/60/c1971934031</span><span>@</span><span>github</span><span>.</span><span>com&gt;</span></span></p>
      <script type="application/ld+json">[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971934031",
"url": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1971934031",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]</script>
    </blockquote>
    <br>
  </body>
</html>

--------------vLrwj5qd3UAOZyER8PRUjCWY--

Namburger commented:
@feranick - thanks :1st_place_medal:

feranick commented:
This is awesome, thanks, @namburger!

On 2/29/24 3:56 PM, Nam Vu wrote:
>
> Merged #60 <https://github.com/google-coral/libedgetpu/pull/60> into 
> master.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/google-coral/libedgetpu/pull/60#event-11971401038>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAAIY7YRAUOCIBTQII7NSIDYV6KXLAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJRHE3TCNBQGEYDGOA>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

--------------GWqrVqHb8HjjQmCgVjOrSwWG
Content-Type: text/html; charset=UTF-8
Content-Transfer-Encoding: 8bit

<!DOCTYPE html><html><head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  </head>
  <body>
    This is awesome, thanks, @namburger!<br>
    <br>
    <div class="moz-cite-prefix">On 2/29/24 3:56 PM, Nam Vu wrote:<br>
    </div>
    <blockquote type="cite" ***@***.***">
      
      <p dir="auto">Merged <a class="issue-link js-issue-link" data-error-text="Failed to load title" data-id="2110560631" data-permission-text="Title is private" data-url="https://github.com/google-coral/libedgetpu/issues/60" data-hovercard-type="pull_request" data-hovercard-url="/google-coral/libedgetpu/pull/60/hovercard" href="https://github.com/google-coral/libedgetpu/pull/60" moz-do-not-send="true">#60</a> into master.</p>
      <p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">—<br>
        Reply to this email directly, <a href="https://github.com/google-coral/libedgetpu/pull/60#event-11971401038" moz-do-not-send="true">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AAAIY7YRAUOCIBTQII7NSIDYV6KXLAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJRHE3TCNBQGEYDGOA" moz-do-not-send="true">unsubscribe</a>.<br>
        You are receiving this because you were mentioned.<img src="https://github.com/notifications/beacon/AAAIY77ZXNTA5GNJWCR6VGTYV6KXLA5CNFSM6AAAAABCTNMRJ2WGG33NNVSW45C7OR4XAZNWJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XKUY3PNVWWK3TUL5UWJTYAAAAAFSMNCVHA.gif" alt="" moz-do-not-send="true" width="1" height="1"><span style="color: transparent; font-size: 0; display: none; visibility: hidden; overflow: hidden; opacity: 0; width: 0; height: 0; max-width: 0; max-height: 0; mso-hide: all">Message
          ID: <span>&lt;google-coral/libedgetpu/pull/60/issue_event/11971401038</span><span>@</span><span>github</span><span>.</span><span>com&gt;</span></span></p>
      <script type="application/ld+json">[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/google-coral/libedgetpu/pull/60#event-11971401038",
"url": "https://github.com/google-coral/libedgetpu/pull/60#event-11971401038",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]</script>
    </blockquote>
    <br>
  </body>
</html>

--------------GWqrVqHb8HjjQmCgVjOrSwWG--

feranick commented:
@namburger, Google used to provide compiled binaries for this lib. While I made several for Linux/OS, I would hope that these will be prepared to be delivered through:
```
https://packages.cloud.google.com/apt
```
as the official channel (and how most people get theirs, along with Windows, etc). Is there any plan for it?

feranick commented:
> @feranick I have recently written a guide on how to get the usb accelerator working with ultralytics yolov8 based on your updated libs which can be found [here](https://docs.ultralytics.com/guides/coral-edge-tpu-on-raspberry-pi/). This may be helpful for people trying to get the tpu working.

In case it's helpful both libcoral and pycoral libraries are now working for me in each of the forked repo. I hope the will be merged after TF2.16.0 reaches stable. Details here:
https://github.com/google-coral/pycoral/issues/137#issuecomment-1974048543

feranick commented:
(PSA: updated libcoral and pycoral libraries are coming). Details here: 

https://github.com/google-coral/pycoral/issues/137#issuecomment-1974048543

Namburger commented:
@feranick looks like https://github.com/tensorflow/tensorflow/pull/63075 is approved!

feranick commented:
That is awesome, @namburger, thanks! Any chance that will be back ported to 2.16.0?On Mar 2, 2024 9:22 PM, Nam Vu ***@***.***> wrote:
@feranick looks like tensorflow/tensorflow#63075 is approved!

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

Namburger commented:
@feranick I'm not sure what's tensorflow plans is for the next release, and I don't have much visibility/influence, unfortunately.
I think in this instance, we should try to get get the PR merged first, then if it won't be ported to a newer release, we can sync this repo on a commit that contains that change, that way we can get all of the coral repos updated asap!

feranick commented:
Thanks, @namburger. This sounds fine on my end.

For reference, a recent PR 
<https://github.com/tensorflow/tensorflow/issues/63018> I submitted was 
pulled by @mihaimaruseac <https://github.com/mihaimaruseac> for TFlite, 
and was quickly backported to 2.16.0. Let's hope this may happen for 
https://github.com/tensorflow/tensorflow/pull/63075 as well.

In any case, I will be here in case you need any action from my side.

It's really awesome, that Coral is slowly getting back on track.

On 3/3/24 2:13 PM, Nam Vu wrote:
>
> @feranick <https://github.com/feranick> I'm not sure what's tensorflow 
> plans is for the next release, and I don't have much 
> visibility/influence, unfortunately.
> I think in this instance, we should try to get get the PR merged 
> first, then if it won't be ported to a newer release, we can sync this 
> repo on a commit that contains that change, that way we can get all of 
> the coral repos updated asap!
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1975268282>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAAIY77GAF5IZ633XT7VBATYWNY5ZAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZVGI3DQMRYGI>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

--------------kxvHDI6JIQRIRYZoi0X3qRfC
Content-Type: text/html; charset=UTF-8
Content-Transfer-Encoding: 8bit

<!DOCTYPE html><html><head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  </head>
  <body>
    Thanks, @namburger. This sounds fine on my end. <br>
    <br>
    For reference, <a href="https://github.com/tensorflow/tensorflow/issues/63018">a
      recent PR</a> I submitted was pulled by <a class="user-mention notranslate" data-hovercard-type="user" data-hovercard-url="/users/mihaimaruseac/hovercard" data-octo-click="hovercard-link-click" data-octo-dimensions="link_type:self" ***@***.***</a> for
    TFlite, and was quickly backported to 2.16.0. Let's hope this may
    happen for <a class="moz-txt-link-freetext" href="https://github.com/tensorflow/tensorflow/pull/63075">https://github.com/tensorflow/tensorflow/pull/63075</a> as
    well. <br>
    <br>
    In any case, I will be here in case you need any action from my
    side.<br>
    <br>
    It's really awesome, that Coral is slowly getting back on track.<br>
    <br>
    <div class="moz-cite-prefix">On 3/3/24 2:13 PM, Nam Vu wrote:<br>
    </div>
    <blockquote type="cite" ***@***.***">
      
      <p dir="auto"><a class="user-mention notranslate" data-hovercard-type="user" data-hovercard-url="/users/feranick/hovercard" data-octo-click="hovercard-link-click" data-octo-dimensions="link_type:self" href="https://github.com/feranick" ***@***.***</a>
        I'm not sure what's tensorflow plans is for the next release,
        and I don't have much visibility/influence, unfortunately.<br>
        I think in this instance, we should try to get get the PR merged
        first, then if it won't be ported to a newer release, we can
        sync this repo on a commit that contains that change, that way
        we can get all of the coral repos updated asap!</p>
      <p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">—<br>
        Reply to this email directly, <a href="https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1975268282" moz-do-not-send="true">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AAAIY77GAF5IZ633XT7VBATYWNY5ZAVCNFSM6AAAAABCTNMRJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZVGI3DQMRYGI" moz-do-not-send="true">unsubscribe</a>.<br>
        You are receiving this because you were mentioned.<img src="https://github.com/notifications/beacon/AAAIY7ZCTXNWG4ZIM6OTKA3YWNY5ZA5CNFSM6AAAAABCTNMRJ2WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTTVXQZ3U.gif" alt="" moz-do-not-send="true" width="1" height="1"><span style="color: transparent; font-size: 0; display: none; visibility: hidden; overflow: hidden; opacity: 0; width: 0; height: 0; max-width: 0; max-height: 0; mso-hide: all">Message
          ID: <span>&lt;google-coral/libedgetpu/pull/60/c1975268282</span><span>@</span><span>github</span><span>.</span><span>com&gt;</span></span></p>
      <script type="application/ld+json">[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1975268282",
"url": "https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1975268282",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]</script>
    </blockquote>
    <br>
  </body>
</html>

--------------kxvHDI6JIQRIRYZoi0X3qRfC--

Namburger commented:
@feranick - I've twisted some knobs in our internal codebase to get [this commit](https://github.com/tensorflow/tensorflow/commit/79ecb3f8bb6bd73f0115fa9a97b630a6f745a426) out, would you like to test now?

feranick commented:
@Namburger Will test as soon as I can today. Quick Q: is this available also on the 2.16.0 branch?

feranick commented:
The reason I am asking is that I am building libcoral against 2.16.0 (as planned) not 2.17.0 as in the current master.

feranick commented:
So I tested it against the current TF master with both MacOS and Linux and all seems to work. To allow the use of TM master for temporary testing, I locally changed this in the libcoral/WORKSPACE`:
```
http_archive(
        name = "org_tensorflow",
        urls = [
            "https://github.com/tensorflow/tensorflow/archive/refs/heads/master.tar.gz",
        ],
        sha256 = "21a8363e3272a19977e2f0d12dcb87d1cb61ff0a79d20cfe456d9840e45e18d6",
        strip_prefix = "tensorflow-" + "master",
        )
```

Namburger commented:
@feranick working with tensorflow is a learning experience for me... I'll do a first pass to see how to get this backported to 2.16, will let you know if we'll need a plan b

feranick commented:
I'd discuss it with @mihaimaruseac. My previous PR was quickly cherrypicked into 2.16.0.

Namburger commented:
@feranick I've also talked to him, doesn't seems like there is a plan to pick this to 2.16 at this time :/

For your change suggestion,
```
http_archive(
        name = "org_tensorflow",
        urls = [
            "https://github.com/tensorflow/tensorflow/archive/refs/heads/master.tar.gz",
        ],
        sha256 = "21a8363e3272a19977e2f0d12dcb87d1cb61ff0a79d20cfe456d9840e45e18d6",
        strip_prefix = "tensorflow-" + "master",
        )
```
Could it works to just set `TENSORFLOW_COMMIT=79ecb3f8bb6bd73f0115fa9a97b630a6f745a426`?

feranick commented:
@Namburger  That is unfortunate. With regards to the specific commit, that would build against 2.17.0 at this point, correct? Given that is unstable and under development, I wouldn't think it's a good idea......

feranick commented:
@Namburger, Compilation of `libcoral` works fine with MacOS and Linux by building against TF commit 79ecb3f. I updated the [PR](https://github.com/google-coral/libcoral/pull/36) to reflect that and now all is working. 

The concerns remains, though, that it's building against an unstable version fo 2.17.0. Also, should be libedgetpu be built against the same TF (stable is now against 2.15.0, but I am prepping the next with TF 2.16 as soon as it reaches stable).

Namburger commented:
@feranick I hear the concerns, it doesn't look like we can get that in for 2.16 so unless we want to sync to a 2.17 dev version, we're stuck in limbo until it comes out.. which probably won't happen any time soon...
It's very unfortunate that ultimately what blocks us is some visibility rules by bazel..

At this point my suggestion to move forward and keep all the main rocal repos updated is to sync to `79ecb3f8bb6bd73f0115fa9a97b630a6f745a426`, we can monitor for breakage and apply updates as 2.17 is being developed. WDYT?

feranick commented:
This sounds good to me @Namburger. The repos are already locally in sync with that commit. 

I already found a possible breakage due to 2.17 [(See log here)](https://github.com/google-coral/libedgetpu/files/14501277/log_arm.txt) when crosscompiling for arm (there are no issues in MacOS and Linux for x86). I forked 2.16.0 and added the visibility patch to test whether the issue is in 2.17 or not. Will report back.

feranick commented:
@namburger since this PR has been merged and concerns libedgetpu (not libcoral), I am going to continue this reporting to the relevant PR:

https://github.com/google-coral/libcoral/pull/36

Namburger commented:
> This sounds good to me @Namburger. The repos are already locally in sync with that commit.
> 
> I already found a possible breakage due to 2.17 [(See log here)](https://github.com/google-coral/libedgetpu/files/14501277/log_arm.txt) when crosscompiling for arm (there are no issues in MacOS and Linux for x86). I forked 2.16.0 and added the visibility patch to test whether the issue is in 2.17 or not. Will report back.

Interesting, doesn't look like `neon_fully_connected_arm32.cc` have been changed for 7 months, that seems to indicates the compiler doesn't like the asm..

mihaimaruseac commented:
So, unfortunately this landed too late in the 2.16 release process (it was supposed to also have an RC1 to handle bugs from RC0 but the release team decided otherwise).

If you pin to a commit from master branch, you will always build at that commit, but will need to always test when you move to a separate one. Regarding the commit to pick, I would suggest either the commit where the support got added or one of the nightly commits afterwards (check if there is a pip wheel built)

feranick commented:
Well, as unfortunate as that is, I am really grateful to you both @mihaimaruseac and @Namburger to try to push this forward. I am still running a few build tests, it's pretty easy to set either version we want to build against. 

If I were to build against the latest nightly, how do I find its corresponding commit?

mihaimaruseac commented:
If you build on the day immediately after the nightly release, you can use the commit at the top of the [nightly branch](https://github.com/tensorflow/tensorflow/tree/nightly)

If you build some days later, `tf.version` should provide the information (I don't recall the exact API, left the actual TF team around 2 years ago and now I'm just consulting from time to time). Alternatively, you can look at the matching [GH Action run](https://github.com/tensorflow/tensorflow/actions/workflows/update-nightly.yml) and take the commit where the action ran at. That would be `019e960` in this screenshot:
![Screenshot from 2024-03-05 14-00-07](https://github.com/google-coral/libedgetpu/assets/323199/cb430f63-bce4-4c75-98c0-b8015e359723)
(which corresponds to https://github.com/tensorflow/tensorflow/actions/runs/8150842615)

feranick commented:
> Interesting, doesn't look like `neon_fully_connected_arm32.cc` have been changed for 7 months, that seems to indicates the compiler doesn't like the asm..

The issue is only specific to arm7a. This is the commit where edits are exactly what the compiler is complaining about. There are no specific details on what drove the commit in first place....

https://github.com/tensorflow/tensorflow/commit/8419c70e2e14b79ad2c514835bf19830050c528e

Namburger commented:
> > Interesting, doesn't look like `neon_fully_connected_arm32.cc` have been changed for 7 months, that seems to indicates the compiler doesn't like the asm..
> 
> The issue is only specific to arm7a. This is the commit where edits are exactly what the compiler is complaining about. There are no specific details on what drove the commit in first place....
> 
> [tensorflow/tensorflow@8419c70](https://github.com/tensorflow/tensorflow/commit/8419c70e2e14b79ad2c514835bf19830050c528e)

humm, that certainly sounds like the issue. As I understand it, that commit changes the constraint of the registers used for that ops which matches the error:
```
external/org_tensorflow/tensorflow/lite/kernels/internal/optimized/4bit/neon_fully_connected_arm32.cc: In function 'void tflite::optimized_4bit::NeonRunKernelNoSDot(const uint8_t*, const int8_t*, int32_t*, int, int, int, int, int, int) [with int RowsLeft = 4; int RowsRight = 1; int Cols = 32]':
external/org_tensorflow/tensorflow/lite/kernels/internal/optimized/4bit/neon_fully_connected_arm32.cc:192:7: error: 'asm' operand has impossible constraints
  192 |       asm volatile(KERNEL_4x1
      |       
```

feranick commented:
Indeed. Weirdly enough, the change in that commit was only applied to 
```
neon_fully_connected_arm32.cc
```
not on 
```
neon_fully_connected_aarch64_sdot.cc
```

which in fact works fine....

feranick commented:
@namburger would you re commend opening an issue?On Mar 5, 2024 6:20 PM, Nam Vu ***@***.***> wrote:


Interesting, doesn't look like neon_fully_connected_arm32.cc have been changed for 7 months, that seems to indicates the compiler doesn't like the asm..

The issue is only specific to arm7a. This is the commit where edits are exactly what the compiler is complaining about. There are no specific details on what drove the commit in first place....
***@***.***

humm, that certainly sounds like the issue. As I understand it, that commit changes the constraint of the registers used for that ops which matches the error:
external/org_tensorflow/tensorflow/lite/kernels/internal/optimized/4bit/neon_fully_connected_arm32.cc: In function 'void tflite::optimized_4bit::NeonRunKernelNoSDot(const uint8_t*, const int8_t*, int32_t*, int, int, int, int, int, int) [with int RowsLeft = 4; int RowsRight = 1; int Cols = 32]':
external/org_tensorflow/tensorflow/lite/kernels/internal/optimized/4bit/neon_fully_connected_arm32.cc:192:7: error: 'asm' operand has impossible constraints
  192 |       asm volatile(KERNEL_4x1
      |       


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

feranick commented:
I tested it with a local fork of TF 2.17.0 with the `neon_fully_connected_arm32.cc` reverted back to be in line with the other architectures, and compilation still fails even with the same error:
```
external/org_tensorflow/tensorflow/lite/kernels/internal/optimized/4bit/neon_fully_connected_arm32.cc: In function 'void tflite::optimized_4bit::NeonRunKernelNoSDot(const uint8_t*, const int8_t*, int32_t*, int, int, int, int, int, int) [with int RowsLeft = 4; int RowsRight = 1; int Cols = 32]':
external/org_tensorflow/tensorflow/lite/kernels/internal/optimized/4bit/neon_fully_connected_arm32.cc:192:7: error: 'asm' operand has impossible constraints
  192 |       asm volatile(KERNEL_4x1
      |       ^~~
INFO: Elapsed time: 34.950s, Critical Path: 17.60s
```
Note that this file (and the whole `4bit` folder) was introduced with TF 2.14)

So maybe the commit was intended to address this problem, but didn't...

feranick commented:
@Namburger Disclaimer: I know close to nothing about asm. Yet, I tried to address this error for what seems like a limitation in the architecture. Following some comments from [here](https://stackoverflow.com/a/25579843), I replaced "r" with "g" so that in `external/org_tensorflow/tensorflow/lite/kernels/internal/optimized/4bit/neon_fully_connected_arm32.cc` line 192 what is currently:
```
:
                   : [lhs_val] "r"(lhs_val), [rhs_val] "r"(rhs_val),
                     [element_ptr] "r"(element_ptr), [bit_shift] "r"(bit_shift),
                     [run_depth] "r"(run_depth)
```
is changed to:
```
:
                   : [lhs_val] "g"(lhs_val), [rhs_val] "g"(rhs_val),
                     [element_ptr] "g"(element_ptr), [bit_shift] "g"(bit_shift),
                     [run_depth] "g"(run_depth)
```
Compilation proceeds beyind this point, but it then stops with a similar error for another external library:
```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from external/ruy/ruy/pack_arm.cc:16:
external/ruy/ruy/pack_arm.h:492:9: warning: multi-line comment [-Wcomment]
  492 | #endif  // (RUY_PLATFORM_NEON_64 || RUY_PLATFORM_NEON_32) && \
      |         ^
external/ruy/ruy/pack_arm.cc: In function 'void ruy::Pack8bitColMajorForNeon4Cols(const ruy::PackParams8bit&)':
external/ruy/ruy/pack_arm.cc:264:3: error: 'asm' operand has impossible constraints
  264 |   asm volatile(
      |   ^~~
INFO: Elapsed time: 32.683s, Critical Path: 20.21s
```
Not encouraging.

Namburger commented:
hi @feranick with all of the changes, would you be able to send a set of commands to reproduce the very first issues?
I believe it looks something like this? ( but may requires you to push some commits to your folk of tensorflow and then sync your fork of libedgetpu to it?)
```
git clone git@github.com:google-coral/libcoral.git && cd libcoral
git fetch origin pull/36/head:pull-36
git checkout pull-36
git submodule init && git submodule update
make DOCKER_IMAGE=debian:buster DOCKER_CPUS="armv7a" DOCKER_TARGETS=tests docker-build
```
I'm talking to some folks, trying to get an easy process to repro the original issues and continue on. It appears that `element_ptr` needs to be a `r+` and there could be more issues as you pointed out so we'll need to work this out one step at a time, unfortunately :(

Namburger commented:
A procedure to making code changes to tensorflow and test them would be nice to have also, not sure if this can be easily done with bazel, though

mihaimaruseac commented:
One idea would be to fork TensorFlow, make changes in your own fork (and sync from time to time) and then change https://github.com/google-coral/libedgetpu/blob/b5820ad9437548febcf58f593a665541e7c5e561/workspace.bzl#L68-L76 to point to the fork and the last commit there.

Then, once it all works, you can make a PR from the fork back to TF and once that lands you can change the `workspace.bzl` file to point back to upstream TF

feranick commented:
@mihaimaruseac Thanks. This is effectively what I did for testing. An advantage of this is that `ibedgetpu` could still be built on a stable version of TF (2.16.0, forked and with the visibility patch). I'd note that `libedgetpu` can be used independently from libcoral/pycoral (in fact that is my way of using the edgeTPU). Yet it would allow to keep testing and building libcoral/pycoral, while libedgetpu would be "stable".

feranick commented:
> hi @feranick with all of the changes, would you be able to send a set of commands to reproduce the very first issues? I believe it looks something like this? ( but may requires you to push some commits to your folk of tensorflow and then sync your fork of libedgetpu to it?)

Hi @Namburger , here 's a slightly revised version of your commands (I can't clone via git@github.com:google-coral/libcoral.git, only via https). Also you should build using debian:bookworm as it has support for python3.11:

```
git clone https://github.com/google-coral/libcoral.git && cd libcoral
git fetch origin pull/36/head:pull-36
git checkout pull-36
git submodule init && git submodule update
make DOCKER_IMAGE=debian:bookworm DOCKER_CPUS="armv7a" DOCKER_TARGETS=tests docker-build
```

This commands will surely trigger the issue (just tested it). As for `libedgetpu` it currently builds against TF 2.16.0-rc0, but the issue here happens regardless of the TF version libedgetpu is built against. Can change it to TF 2.17.0+visibility if needed (yet, see [comment](https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1981234057))

Namburger commented:
by mean of an update, the author of tensorflow/tensorflow@8419c70 promised to take a look sometimes this week!

feranick commented:
> by mean of an update, the author of [tensorflow/tensorflow@8419c70](https://github.com/tensorflow/tensorflow/commit/8419c70) promised to take a look sometimes this week!

Hi @Namburger, just wondering whether there is an update on this... Thanks!

rushowr commented:
Joining in the fray here, wondering how the future for the Coral TPU is going to look. Many of us out here bought this with hopes of affordable facial and object recog for software like Frigate using the Coral devices....

feranick commented:
To be honest with you, the development is pretty much dead. Google is happy to sell you the boards, ASUS will do so for bulk sales, and yet the platform is dead. Shameful if you are asking me. This his why I embarked in the effort of keeping it alive, and with some success to merge it with the main repo. But I also have a fork for the latest. I am pretty sure that apart from some googler with some time to spare, Google basically killed their dev team. So I would not be invested too much into the platform, unless you are willing to put the  work in.

rushowr commented:
Luckily I only purchase a single pcie Coral. Unfortunately for me, I'm on a fixed income so the loss hurts just a smidge more than it would have if I was still at 6-figure income
